### PR TITLE
routing/http: feat: add streaming support

### PR DESCRIPTION
**Pull request by guseggert**
This adds streaming support to the routing/v1 client and server by changing the interfaces to use iterators instead of slices, and adding content type negotation to the client and server with `application/x-ndjson` support.

### CODE REVIEW

File: routing/http/client/client.go
Comment (ajnavarro) [inline routing/http/client/client.go]: shall this be `mediaTypeNDJSON`?

File: routing/http/types/iter/iter.go
Comment (ajnavarro) [inline routing/http/types/iter/iter.go]: We should add a Close() method to avoid the need of exhausting the iterator to free resources.

File: routing/http/client/client.go
Comment (ajnavarro) [inline routing/http/client/client.go]: We could send a measurement to the iterator and add a counter to measure received elements on close or on finish.

File: routing/http/server/server.go
Comment (ajnavarro) [inline routing/http/server/server.go]: shall we use here the previously defined constants?

File: routing/http/server/server.go
Comment (ajnavarro) [inline routing/http/server/server.go]: the same as the previous comment: Shall we use here the previously defined constant?

File: routing/http/server/server.go
Comment (ajnavarro) [inline routing/http/server/server.go]: shall we stop iteration when error here?

File: routing/http/types/iter/iter.go
Comment (ajnavarro) [inline routing/http/types/iter/iter.go]: WDYT about:

```suggestion
	Next(ctx context.Context) (T, error)
	io.Closer
```
Where instead of using a boolean to know if we still have new elements, we use `io.EOF`? (this just simplify the method definition and also a little bit the code and the implementation IMHO)

**Review by ajnavarro — DISMISSED** [review]

File: routing/http/types/iter/iter.go
Comment (guseggert) [reply routing/http/types/iter/iter.go]: I don't feel strongly about either. I always found `io.EOF` to be clunky since it's not really an error so it's easy to miss or handle incorrectly due to being mislabeled, but OTOH it's still idiomatic and having three return values can also be confusing.  I'll try the EOF and see how it looks.

Comment (guseggert) [reply routing/http/types/iter/iter.go]: I also don't think a ctx is necessary (including in JSONIter where I included one), this is controlled by the underlying reader being closed (e.g. when reading client-side and the req context is canceled, the body reader is closed and the context err is returned).

Comment (guseggert) [reply routing/http/types/iter/iter.go]: I tried it with `io.EOF` and it was pretty annoying to have to do `if errors.Is(err, io.EOF){...}` instead of `if ok {...}`.  Definitely adding `Close()` though.

Comment (ajnavarro) [reply routing/http/types/iter/iter.go]: > I tried it with io.EOF and it was pretty annoying to have to do if errors.Is(err, io.EOF){...} instead of if ok {...}. 

io.EOF is intended to be used like:

```go
if err == io.EOF {
    break
}
if err != nil {
    return err
}

```

I think is quite similar to `if ok {...}` and 100% idiomatic. In any case, the final decision is up to you, I don't want to bikeshed.

File: routing/http/types/ndjson/provider.go
Comment (ajnavarro) [inline routing/http/types/ndjson/provider.go]: Iterator chaining is a good example to see how this can be simplified using only two return params:

```go
v, err := p.iter.Next()
if err != nil {
   return nil, err
}
...
```

File: routing/http/types/ndjson/provider.go
Comment (ajnavarro) [inline routing/http/types/ndjson/provider.go]: Another example how having only two params simplifies iterator implementations and it is less error prone (there is no chance to flip an ok boolean when is not the case):

```go
err := json.Unmarshal(...)
return &prov, err
```

File: routing/http/types/iter/iter.go
Comment (ajnavarro) [reply routing/http/types/iter/iter.go]: Sorry, after reviewing the code, I changed my mind :rofl: . Implementation-wise, using 2 return parameters instead of 3 has several advantages:
- Less error-prone. The developer doesn't have to decide when the returned boolean has to be true or false, it depends only on the error output. 
- Easy to chain iterators, less code doing the same. Check: https://github.com/ipfs/go-libipfs/pull/18#discussion_r1072012310
- Easy to integrate with existing functions. AFAIK there is no example on the Go SDK doing what we are doing with the Iterator, returning an error and an ok flag value simultaneously. That makes integration with existing APIS more verbose. See example here: https://github.com/ipfs/go-libipfs/pull/18#discussion_r1072015158

Comment (guseggert) [reply routing/http/types/iter/iter.go]: > Less error-prone. The developer doesn't have to decide when the returned boolean has to be true or false, it depends only on the error output.

I disagree with this characterization, because you have to remember to check io.EOF. Typically I associate that with io.Reader, not other random non-io fns.

The other points are ultimately about ease of use, which is a secondary concern to me. Iterating over a list of stuff that does preemptible I/O, and could error out at any point, is not easy, and the complex interface reflects that instead of hiding it behind a sentinel error.

File: routing/http/types/ndjson/provider.go
Comment (guseggert) [reply routing/http/types/ndjson/provider.go]: btw this is an example of why you should use `errors.Is(err, io.EOF)`, I almost always wrap these kinds of errors with some context which would break `err == io.EOF` checks

Comment (ajnavarro) [reply routing/http/types/ndjson/provider.go]: `io.EOF` is intended to be used like `err == io.EOF` and must not be wrapped because it is a sentinel error: https://github.com/golang/go/issues/39155

File: routing/http/types/iter/iter.go
Comment (ajnavarro) [reply routing/http/types/iter/iter.go]: > I disagree with this characterization, because you have to remember to check io.EOF

This is the same as saying that you have to remember to check the flag... you have to check something to know when to stop iterating. Using a sentinel error makes to only have one place to check. Errors that need to be returned by an interface must be described on the Godocs methods, that's how people will know how to check when the iterator is finished.

My point is that this Iterator API is like IO API, if it returns an error, then you must stop reading, so having a boolean flag duplicates that information. Every time an error is returned on an iterator implementation, the flag will be false.

That's why I think a sentinel error like io.EOF is a good idea here: we simplify the implementations, we simplify the interaction with other standard libraries, and we avoid duplicated method outputs.

File: routing/http/server/server.go
Comment (masih) [inline routing/http/server/server.go]: Check for `*/*`?

File: routing/http/types/iter/iter.go
Comment (ajnavarro) [reply routing/http/types/iter/iter.go]: @guseggert WDYT about https://pkg.go.dev/database/sql#Rows approach?

Interface:
```go
type Iter[T any] interface {
    Next() bool
    Err() error
    Value() T
    Close() error
}
```

Example:
```go
if iter.Next() {
    val := iter.Value()
    ....
}

if err := iter.Err(); err != nil {
    panic(err)
}

if err := iter.Close(); err != nil {
    panic(err)
}
```

No io.EOF check, no 3 return values, we are following the standard API approach

Comment (guseggert) [reply routing/http/types/iter/iter.go]: sgtm thanks for finding this

File: routing/http/types/ndjson/provider.go
Comment (guseggert) [reply routing/http/types/ndjson/provider.go]: even worse then that you cant use standard idioms with it

Comment (Jorropo) [reply routing/http/types/ndjson/provider.go]: > btw this is an example of why you should use `errors.Is(err, io.EOF)`, I almost always wrap these kinds of errors with some context which would break `err == io.EOF` checks

It is widely accepted that `io.EOF` must not be wrapped:
> EOF is the error returned by Read when no more input is available. (Read must return EOF itself, not an error wrapping EOF, because callers will test for EOF using ==.) Functions should return EOF only to signal a graceful end of input. If the EOF occurs unexpectedly in a structured data stream, the appropriate error is either ErrUnexpectedEOF or some other error giving more detail.

Comment (guseggert) [reply routing/http/types/ndjson/provider.go]: > EOF is the error returned by Read when no more input is available

This is not Read()...

**Comment by BigLep** [comment]
@guseggert : what's remaining so this can move out of draft?

File: routing/http/client/client.go
Comment (aschmahmann) [inline routing/http/client/client.go:167]: Is this still a todo?

File: routing/http/client/client.go
Comment (aschmahmann) [inline routing/http/client/client.go:144]: switching to calling `Body.Close()` everywhere is fine, but feels a bit easy to miss. Wondering if something like this might be better:
```
defer func() { 
  if err != nil || isSyncCall {
      resp.Body.Close()
  }
}
```

File: routing/http/types/iter/map.go
Comment (aschmahmann) [inline routing/http/types/iter/map.go:4]: Unused code?

File: routing/http/client/client.go
Comment (guseggert) [reply routing/http/client/client.go:167]: yes, nothing has changed here I'm just making a note that the measurements aren't tested

**Comment by BigLep** [comment]
2023-05-09 maintainer conversation: 
- We want to land this before @guseggert departs
- [x] @guseggert rebase
- @hacdias will review
- [x] @guseggert confirm that this can be bubbled up to Kubo and the cid.contact calls still work (verify it works in production)
- Leave some notes in https://github.com/ipfs/specs/issues/344.  Notes around error handling for example.

**Comment by guseggert** [comment]
Re: testing with cid.contact, I talked with @masih and I was wrong--this is not implemented in cid.contact yet. It is for IPNI APIs but not for routing/v1. Confirmed by wiring into Kubo, req Accept=application/x-ndjson,application/json but resp Content-Type=application/json.  Tracking issue for this on their side: https://github.com/ipni/indexstar/issues/92

I think the best option is to just go ahead and ship this in Kubo and let cid.contact own the integration testing when they are ready to deploy.

**Review by hacdias — APPROVED** [review]
LGTM, thanks! I also went over previous comments to check if everything was addressed. Let's finally merge this!

**Pull request merged**
